content/blog/_posts/2021-04-rust-custom-test-harness.md

```diff
@@ -0,0 +1,296 @@
+}
+
+inventory::submit!(IntegrationTest {
+   name: "basic",
+   test_fn: basic_test
+});
+```
+
+We use `inventory::submit!()` to register our new test with the `IntegrationTest` struct we defined earlier.
+
+`name` is a friendly, human-readable name. We can use this name as a key to search through the plugin registry.
+
+`test_fn` takes the name of our test function. It has the same function signature as we defined. 
+
+
+### Running tests from registry
+
+We’ll finish this example up by running all of our registered tests
+
+```diff
+// main.rs
+
+pub mod tests;
++ use tests::IntegrationTest;
+
+fn setup() {
+   println!("Setup")
+}
+
+fn teardown() {
+   println!("Teardown")
+}
+fn main() {
+   setup();
+
+-   // TODO: Run the test
++   // Run the tests
++   for t in inventory::iter::<IntegrationTest> {
++       (t.test_fn)()
++   }
+
+   teardown();
+}
+```
+
+```shell
+$ cargo test integration
+   Compiling blog-post-example v0.1.0 (/home/telant/Documents/blog-post-example)
+	Finished test [unoptimized + debuginfo] target(s) in 0.21s
+ 	Running target/debug/deps/blog_post_example-e042d787684bb333
+
+running 0 tests
+
+test result: ok. 0 passed; 0 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.00s
+
+ 	Running target/debug/deps/integration-7ed2452642c6f3b6
+
+Setup
+Running basic test
+Teardown
+```
+
+### Tips for extending the example
+
+The example runs all of the registered tests. But here are some useful impls if you want to extend even further. For example, adding a CLI, if you want to select individual tests. Or provide options to customize setup or teardown behavior.
+
+```rust
+impl IntegrationTest {
+   pub fn all_test_names() -> Vec<&'static str> {
+       inventory::iter::<IntegrationTest>
+           .into_iter()
+           .map(|x| x.name)
+           .collect::<Vec<&str>>()
+   }
+
+   pub fn from_name<S: AsRef<str>>(test_name: S) -> Option<&'static IntegrationTest> {
+       inventory::iter::<IntegrationTest>
+           .into_iter()
+           .find(|t| t.name == test_name.as_ref())
+   }
+}
+```
+
+If you want to see more of these ideas extended even further for a complex system, check out [Fluvio’s integration test runner](https://github.com/infinyon/fluvio/tree/master/tests/runner/src). We use a CLI and async functions, and an attribute procedural macro to cut down on some of the boilerplate.
+
+### Conclusion
+
+Rust’s testing ecosystem in the 2018 edition is great for unit testing. But for integration testing it still has room for improvement. Custom harnesses will become more necessary as Rust gains new developers. 
+
+If we want to avoid reinventing the wheel, we need stable support from [libtest](https://docs.rs/libtest/0.0.1/libtest/) or more examples of how to perform test collection and patterns for setup, test, teardown workflows.
+
+If you made it this far, thank you! I wish I had a guide like this before setting out on this development journey. Hopefully others find my experience helpful.
```

Comment:
I think it is good to plug Fluvio at the end of the blog, I usually add a "Quick links" section to take people to other significant pages

Reply:
I usually reserve quick links for reference material, or related posts. Posts tend to feel older sooner if the footer links are to old and irrelevant content just to self-advertise.

If we want to hype other posts in that manner, it should be built into the hugo template.